pkg/ccl/backupccl/memory_utils.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2017 The Cockroach Authors.
+//
+// Licensed as a CockroachDB Enterprise file under the Cockroach Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package backupccl
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/util/mon"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+)
+
+// memoryAccumulator is a thin wrapper around a BoundAccount that only releases memory
+// from the bound account when it is closed, otherwise it accumulates and
+// re-uses resources.
+// This is useful when resources, once accumulated should not be returned as
+// they may be needed later to make progress.
+// It is safe for concurrent use.
+type memoryAccumulator struct {
+	syncutil.Mutex
+	ba       mon.BoundAccount
+	reserved int64
+}
+
+// newMemoryAccumulator creates a new accumulator backed by a bound account created
+// from the given memory monitor.
+func newMemoryAccumulator(mm *mon.BytesMonitor) *memoryAccumulator {
+	if mm == nil {
+		return nil
+	}
+	return &memoryAccumulator{ba: mm.MakeBoundAccount()}
+}
+
+// request checks that the given number of bytes is available, requesting some
+// from the backing monitor if necessary.
+func (acc *memoryAccumulator) request(ctx context.Context, requested int64) error {
+	if acc == nil {
+		return nil
+	}
+
+	acc.Lock()
+	defer acc.Unlock()
+
+	if acc.reserved >= requested {
+		acc.reserved -= requested
+		return nil
+	}
+
```

Comment:
should we reduce requested by reserved then zero out reserved here before we go then ask for the net from ba?

Reply:
yup, good idea, done.